drivers/network/dd/b57xx/hardware.c

```diff
@@ -0,0 +1,1582 @@
+        if (!(Value32 & B57XX_ETH_MI_COMMUNICATION_START))
+        {
+            break;
+        }
+        NdisStallExecution(1);
+    }
+    if (Value32 & B57XX_ETH_MI_COMMUNICATION_START)
+    {
+        NDIS_MinDbgPrint("Unable to write to PHY, timed out (0x%x)\n", Value32);
+        return NDIS_STATUS_FAILURE;
+    }
+    
+    return NDIS_STATUS_SUCCESS;
+}
+
+VOID
+B57XXClearMACAttentions(_In_ PB57XX_ADAPTER Adapter)
+{
+    ULONG Value;
+    
+    B57XXEnableRegister(Adapter,
+                        B57XX_REG_ETH_MAC_STATUS,
+                        &Value,
+                        B57XX_ETH_MAC_STATUS_CONFIGCHANGED |
+                        B57XX_ETH_MAC_STATUS_SYNCCHANGED |
+                        B57XX_ETH_MAC_STATUS_MICOMPLETE);
+}
+
+VOID
+B57XXReadAndClearPHYInterrupts(_In_  PB57XX_ADAPTER Adapter,
+                               _Out_ PUSHORT Interrupts)
+{
+    USHORT Dummy;
+    
+    B57XXReadPHY(Adapter, B57XX_MII_REG_INT_STATUS, Interrupts);
+    // Read again to clear
+    B57XXReadPHY(Adapter, B57XX_MII_REG_INT_STATUS, &Dummy);
+}
+
+VOID
+B57XXReadStatistic(_In_  PB57XX_ADAPTER Adapter,
+                   _In_  ULONG BlockOffset,
+                   _In_  ULONG RegOffset,
+                   _In_  BOOLEAN Is64Bit,
+                   _Out_ PULONG64 pValue64)
+{
+    PULARGE_INTEGER pLargeInt = (PULARGE_INTEGER)pValue64;
+    
+    if (!Adapter->B5705Plus)
+    {
+        if (Is64Bit)
+        {
+            pLargeInt->HighPart = *(PULONG)((PUCHAR)Adapter->Statistics.pBlock + BlockOffset + 0);
+            pLargeInt->LowPart = *(PULONG)((PUCHAR)Adapter->Statistics.pBlock + BlockOffset + 4);
+        }
+        else
+        {
+            pLargeInt->LowPart = *(PULONG)((PUCHAR)Adapter->Statistics.pBlock + BlockOffset);
+        }
+    }
+    else
+    {
+        if (Is64Bit)
+        {
+            B57XXReadULong(Adapter, RegOffset + 0, &pLargeInt->HighPart);
+            B57XXReadULong(Adapter, RegOffset + 4, &pLargeInt->LowPart);
+        }
+        else
+        {
+            B57XXReadULong(Adapter, RegOffset, &pLargeInt->LowPart);
+        }
+    }
+}
+
+VOID
+B57XXAccumulateAddress(_Out_ PB57XX_PHYSICAL_ADDRESS pDestination,
+                       _In_ LONG Offset)
+{
+    NDIS_PHYSICAL_ADDRESS Address;
+    
+    Address.HighPart = pDestination->HighPart;
+    Address.LowPart = pDestination->LowPart;
+    
+    Address.QuadPart += Offset;
+    
+    pDestination->HighPart = Address.HighPart;
+    pDestination->LowPart = Address.LowPart;
+}
+
+ULONG
+B57XXComputeInverseCrc32(_In_ PUCHAR pBuffer,
+                         _In_ ULONG BufferSize)
+{
+    ULONG Tmp;
+    ULONG Reg = 0xFFFFFFFF;
+    
+    for(ULONG i = 0; i < BufferSize; i++)
+    {
+        Reg ^= pBuffer[i];
+        
+        for(ULONG j = 0; j < 8; j++)
+        {
+            Tmp = Reg & 0x01;
+            Reg >>= 1;
+            if(Tmp)
+            {
+                Reg ^= 0xEDB88320;
+            }
```

Comment:
This could be optimized:
```suggestion
            Reg = (Reg >> 1) ^ (-(LONG)(Reg & 1) & 0xEDB88320);
```
(a lookup table would be overkill for multicast hashing)